src/Effects/CombFilterBank.cpp

```diff
@@ -0,0 +1,128 @@
+    baseFreq(110.0f),
+    nrOfStrings(0),
+    memory(*alloc),
+    samplerate(samplerate_),
+    buffersize(buffersize_)
+    {
+        // setup the smoother for gain parameter
+        gain_smoothing.sample_rate(samplerate/16);
+        gain_smoothing.thresh(0.02f); // TBD: 2% jump audible?
+        gain_smoothing.cutoff(1.0f);
+        pos_writer = 0;
+    }
+
+    CombFilterBank::~CombFilterBank()
+    {
+        setStrings(0, baseFreq);
+    }
+
+    void CombFilterBank::setStrings(unsigned int nrOfStringsNew, const float baseFreqNew)
+    {
+        // limit nrOfStringsNew
+        nrOfStringsNew = min(NUM_SYMPATHETIC_STRINGS,nrOfStringsNew);
+
+        if(nrOfStringsNew == nrOfStrings && baseFreqNew == baseFreq)
+            return;
+
+        const unsigned int mem_size_new = (int)ceilf(( (float)samplerate/baseFreqNew*1.03f + buffersize + 2)/16) * 16;
+        if(mem_size_new == mem_size)
+        {
+            if(nrOfStringsNew>nrOfStrings)
+            {
+                for(unsigned int i = nrOfStrings; i < nrOfStringsNew; ++i)
+                {
+                    string_smps[i] = memory.valloc<float>(mem_size);
+                    memset(string_smps[i], 0, mem_size*sizeof(float));
+                }
+            }
+            else if(nrOfStringsNew<nrOfStrings)
+                for(unsigned int i = nrOfStringsNew; i < nrOfStrings; ++i)
+                    memory.devalloc(string_smps[i]);
+        } else
+        {
+            // free the old buffers (wrong size for baseFreqNew)
+            for(unsigned int i = 0; i < nrOfStrings; ++i)
+                memory.devalloc(string_smps[i]);
+
+            // allocate buffers with new size
+            for(unsigned int i = 0; i < nrOfStringsNew; ++i)
+            {
+                string_smps[i] = memory.valloc<float>(mem_size_new);
+                memset(string_smps[i], 0, mem_size_new*sizeof(float));
+            }
+            // update mem_size and baseFreq
+            mem_size = mem_size_new;
+            baseFreq = baseFreqNew;
+        }
+        // update nrOfStrings
+        nrOfStrings = nrOfStringsNew;
+    }
+
+    inline float CombFilterBank::tanhX(const float x)
+    {
+        // Pade approximation of tanh(x) bound to [-1 .. +1]
+        // https://mathr.co.uk/blog/2017-09-06_approximating_hyperbolic_tangent.html
+        const float x2 = x*x;
+        return (x*(105.0f+10.0f*x2)/(105.0f+(45.0f+x2)*x2));
+    }
+
+    inline float CombFilterBank::sampleLerp(const float *smp, const float pos) {
+        int poshi = (int)pos; // integer part (pos >= 0)
+        float poslo = pos - (float) poshi; // decimal part
+        // linear interpolation between samples
+        return smp[poshi] + poslo * (smp[poshi+1]-smp[poshi]);
+    }
+
+    void CombFilterBank::filterout(float *smp)
+    {
+        // no string -> no sound
+        if (nrOfStrings==0) return;
+
+        // interpolate gainbuf values over buffer length using value smoothing filter (lp)
```

Comment:
Do you mind adding a comment that this interpolates from 0 to `gainbwd`? Or (maybe even better?) add it in `DSP/Value_Smoothing_Filter.h`? At least I had to add debug printfs to understand it :grin:

Reply:
it does not interpolate from 0. it's just a (up-sampling) lp filter for parameter values.